crates/shared/src/price_estimation/sanitized.rs

```diff
@@ -0,0 +1,245 @@
+use super::{ensure_token_supported, Estimate, PriceEstimating, PriceEstimationError, Query};
+use crate::bad_token::BadTokenDetecting;
+use anyhow::Result;
+use futures::future;
+use model::order::BUY_ETH_ADDRESS;
+use primitive_types::{H160, U256};
+use std::sync::Arc;
+
+const ETH_UNWRAP_COST: U256 = U256([0, 0, 0, 123]);
+
+/// Verifies that buy and sell tokens are supported and handles
+/// ETH as buy token appropriately.
+pub struct SanitizedPriceEstimator<T: PriceEstimating> {
+    inner: T,
+    bad_token_detector: Arc<dyn BadTokenDetecting>,
+    native_token: H160,
+}
+
+impl<T: PriceEstimating> SanitizedPriceEstimator<T> {
+    pub fn new(
+        inner: T,
+        native_token: H160,
+        bad_token_detector: Arc<dyn BadTokenDetecting>,
+    ) -> Self {
+        Self {
+            inner,
+            native_token,
+            bad_token_detector,
+        }
+    }
+    async fn estimate_sanitized_query(
+        &self,
+        query: &Query,
+    ) -> Result<Estimate, PriceEstimationError> {
+        if query.buy_token == query.sell_token {
+            return Ok(Estimate {
+                out_amount: query.in_amount,
+                gas: 0.into(),
+            });
+        }
+
+        ensure_token_supported(query.buy_token, self.bad_token_detector.as_ref()).await?;
+        ensure_token_supported(query.sell_token, self.bad_token_detector.as_ref()).await?;
+
+        let buy_eth = query.buy_token == BUY_ETH_ADDRESS;
+        let sanitized_query = if buy_eth {
+            Query {
+                buy_token: self.native_token,
+                ..*query
+            }
+        } else {
+            *query
+        };
+
+        let mut estimated_price = self.inner.estimate(&sanitized_query).await?;
+
+        if buy_eth {
+            estimated_price.gas =
+                estimated_price
+                    .gas
+                    .checked_add(ETH_UNWRAP_COST)
+                    .ok_or(anyhow::anyhow!(
+                        "cost of unwrapping ETH would overflow gas price"
+                    ))?;
+        }
+
+        Ok(estimated_price)
+    }
+}
+
+#[async_trait::async_trait]
+impl<T: PriceEstimating> PriceEstimating for SanitizedPriceEstimator<T> {
+    async fn estimates(&self, queries: &[Query]) -> Vec<Result<Estimate, PriceEstimationError>> {
+        future::join_all(
+            queries
+                .iter()
+                .map(|query| self.estimate_sanitized_query(query)),
+        )
+        .await
```

Comment:
nit: The `BaselinePriceEstimator` takes advantage of bundling queries into a single `estimates` "bundle" to parallelize things like fetching pool data using batch `eth_calls` (since nodes handle requests one at a time and not in parallel). With that in mind, it means when processing multiple queries, it is better to send all the `Query`-ies in a single `estimates([...])` call instead of splitting and estimating concurrently many `estimate(...)` calls (because the limiting factor - the node - doesn't process call concurrently very well).

Maybe we should take a similar approach to the `Baseline` solver and first query all the unsupported tokens:

https://github.com/gnosis/gp-v2-services/blob/56f3fa6a16140037fc07e759ec2c6981e551e8b7/crates/shared/src/price_estimation/baseline.rs#L86-L102

And then do the `buy_eth` dance for the remaining valid queries.

Reply:
If this is way too complicated, I'm open to keeping it implemented as it is here, but it would be nice to at least get an idea if it affects our `POST /api/v1/quote` request times (I wouldn't be surprised if it didn't because I don't think that the `Baseline` price estimator is the slowest one, and the slower 0x and ParaSwap estimators don't benefit from this).

But we should at least measure 😄.

Reply:
That sounds absolutely reasonable. I had a similar idea where I wanted to see whether it was possible to use `PriceEstimating::estimates` instead of `PriceEstimating::estimate`. I imagine the implementation of estimating a batch could easily be more efficient than estimating a single query.
I will play around with those ideas and hope they don't make the code too complex.

Reply:
Also I would think that the set of supported tokens wouldn't change during the runtime of the service.
In that case we can keep a cache of unsupported tokens and only query tokens we have never seen before.
What do you think about that?

Reply:
Nevermind. The signature `async fn estimates(&self, queries: &[Query]) -> Vec<Result<Estimate, PriceEstimationError>>` doesn't allow mutating the state of `self` during the estimation and introducing something like a global `RwLock<Cache>` just for that doesn't seem like a great idea.

Reply:
> Also I would think that the set of supported tokens wouldn't change during the runtime of the service.

Unfortunately, that isn't true. Some tokens can be paused and resumed.

Reply:
I was able to implement both ideas but I think the code suffered quite a bit.
1) detecting bad tokens in bulk lead to very weird error handling where a single failure determining the quality of a token leads to the whole batch being dropped. (I took that behavior from the baseline estimator).
2) estimating prices in bulk lead to this weird pattern where I split trivial estimations from estimations required to be forwarded to the wrapped estimator. Merging those 2 back together while preserving the order of the results is not trivial to understand IMO.

If there is some Rust syntax sugar or clever iterator magic we can use to make those changes easier to understand, let me know.

Reply:
> If there is some Rust syntax sugar or clever iterator magic

I suggest including the `EstimationResult` as part of `EstimationProgress` variants that use it.

> Merging those 2 back together while preserving the order of the results is not trivial to understand IMO.

I think its not too complex. Maybe ☝️ helps.

Reply:
As for:

> detecting bad tokens in bulk lead to very weird error handling where a single failure determining the quality of a token leads to the whole batch being dropped. (I took that behavior from the baseline estimator).

Hmm... I'm OK with this because it was the behaviour of `Baseline` solver. The only thing I can think of is storing `Result`s of the `self.bad_token_detector.detect(token).await` call in a `HashMap<H160, Result<BadTokenResult>>` instead of a `HashSet<H160>` of bad tokens, and then checking for each token in `bulk_estimate` if the result was successful (propagating errors ofc.) and if its "good" (returning an "unsupported token" error if not).

Reply:
My idea turned out to not work 😛. I also don't have a better idea then.

Reply:
I think the error handling improved a bit. Instead of `HashMap<H160, Result<BadTokenResult>>` I went for `HashMap<H160, PriceEstimationError>` and converted the different errors directly in `get_token_quality_errors()`. This lead to the early returns in `bulk_estimate_prices()` being less verbose in case of an error. I'm somewhat happy with that solution.

Reply:
Nice solution!